network/transport/grpc/connection_manager.go

```diff
@@ -213,16 +264,103 @@ func (s *grpcConnectionManager) Stop() {
+						cp.dialing.Store(false) // reset call lock at the end of dialing
+						dialingWG.Done()
+					}()
+					s.dial(cp)
+				}(c)
+			}
+		}
+	}
+	dialingWG.Wait()
+}
+
+func (s *grpcConnectionManager) dial(contact *contact) {
+	connection, isNew := s.connections.getOrRegister(s.ctx, contact.peer, true)
 	if !isNew {
-		log.Logger().
-			WithField(core.LogFieldPeerAddr, peer.Address).
-			Info("Connection for peer already exists.")
+		// can only occur when receiving an inbound connection at the same time.
+		log.Logger().WithFields(contact.peer.ToFields()).
+			Debug("stop dialing, already has a connection")
+		return
+	}
+	defer func() {
+		// connection does not exist outside the dialer
+		connection.disconnect()
+		s.connections.remove(connection)
+	}()
+
+	// Open a grpc.ClientConn
+	log.Logger().WithFields(contact.peer.ToFields()).Debug("connecting to peer")
+	contact.attempts.Add(1)
+	contact.lastAttempt.Store(time.Now())
+	dialContext, cancel := context.WithTimeout(s.ctx, s.connectionTimeout)
+	defer cancel()
+	grpcClient, err := s.dialer(dialContext, contact.peer.Address, s.dialOptions...)
+	if err != nil { // failed to connect
+		log.Logger().WithError(err).WithFields(contact.peer.ToFields()).Debug("failed to open a grpc ClientConn")
+		errStatus, isStatusError := status.FromError(err)
+		if isStatusError && errStatus.Code() == codes.Canceled {
+			// Do not backoff when context is cancelled
+			// Backoff might try to persist after stores are closed
+			// https://github.com/nuts-foundation/nuts-node/issues/1864
+			return
+		}
+		contact.backoff.Backoff() // backoff store
 		return
 	}
-	s.startTracking(peer.Address, connection)
+	defer grpcClient.Close()
+	log.Logger().WithFields(contact.peer.ToFields()).Debug("connected to peer (outbound)")
+
+	// Connect protocol streams
+	err = s.openOutboundStreams(connection, grpcClient) // blocking call, dial needs to be async
+	if err != nil {
+		// connection failed, increase backoff
+		// TODO: check if this works as intended for multiple streams/protocols on the same connection
+		contact.backoff.Backoff()
+		if errors.Is(err, ErrUnexpectedNodeDID) {
+			contact.backoff.Reset(time.Hour * 24 * 365 * 100) // backoff expires after 100 years. Backoff must be reset when peer's (DID) address is updated.
+		}
+		log.Logger().WithError(err).WithFields(connection.Peer().ToFields()).
+			Debug("Error while setting up outbound gRPC streams, disconnecting")
+	} else {
+		// Connection was OK, but now disconnected. Add a random wait to prevent simultaneous reconnecting.
+		contact.backoff.Reset(RandomBackoff(time.Second, 5*time.Second))
+	}
+}
+
+func (s *grpcConnectionManager) hasActiveConnection(peer transport.Peer) bool {
+	if peer.NodeDID.Empty() { // bootstrap matches on address + empty node DID
+		return s.connections.Get(ByAddress(peer.Address), ByNodeDID(did.DID{})) != nil
+	}
+	// Only authenticated connections
+	return s.connections.Get(ByNodeDID(peer.NodeDID), ByAuthenticated()) != nil
```

Comment:
It is not an authenticated connection yet and might not become one. ByAuthenticated() will always return true though since inbound does not create a connection until the stream is verified (and thus authenticated) and outbound will not allow 2 simultaneous calls